# Car Price Prediction

## Data Cleaning

In [1]:
#Writig the libraries
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings("ignore")

In [2]:
#Showing the DataSet
df = pd.read_csv(r'C:\Datos\archive\Car details v3.csv')
df

name  year  selling_price  km_driven    fuel  \
0           Maruti Swift Dzire VDI  2014         450000     145500  Diesel   
1     Skoda Rapid 1.5 TDI Ambition  2014         370000     120000  Diesel   
2         Honda City 2017-2020 EXi  2006         158000     140000  Petrol   
3        Hyundai i20 Sportz Diesel  2010         225000     127000  Diesel   
4           Maruti Swift VXI BSIII  2007         130000     120000  Petrol   
...                            ...   ...            ...        ...     ...   
8123             Hyundai i20 Magna  2013         320000     110000  Petrol   
8124         Hyundai Verna CRDi SX  2007         135000     119000  Diesel   
8125        Maruti Swift Dzire ZDi  2009         382000     120000  Diesel   
8126               Tata Indigo CR4  2013         290000      25000  Diesel   
8127               Tata Indigo CR4  2013         290000      25000  Diesel   

     seller_type transmission                 owner     mileage   engine  \
0     Individual       Manual           First Owner   23.4 kmpl  1248 CC   
1     Individual       Manual          Second Owner  21.14 kmpl  1498 CC   
2     Individual       Manual           Third Owner   17.7 kmpl  1497 CC   
3     Individual       Manual           First Owner   23.0 kmpl  1396 CC   
4     Individual       Manual           First Owner   16.1 kmpl  1298 CC   
...          ...          ...                   ...         ...      ...   
8123  Individual       Manual           First Owner   18.5 kmpl  1197 CC   
8124  Individual       Manual  Fourth & Above Owner   16.8 kmpl  1493 CC   
8125  Individual       Manual           First Owner   19.3 kmpl  1248 CC   
8126  Individual       Manual           First Owner  23.57 kmpl  1396 CC   
8127  Individual       Manual           First Owner  23.57 kmpl  1396 CC   

       max_power                     torque  seats  
0         74 bhp             190Nm@ 2000rpm    5.0  
1     103.52 bhp        250Nm@ 1500-2500rpm    5.0  
2         78 bhp      12.7@ 2,700(kgm@ rpm)    5.0  
3         90 bhp   22.4 kgm at 1750-2750rpm    5.0  
4       88.2 bhp      11.5@ 4,500(kgm@ rpm)    5.0  
...          ...                        ...    ...  
8123   82.85 bhp           113.7Nm@ 4000rpm    5.0  
8124     110 bhp  24@ 1,900-2,750(kgm@ rpm)    5.0  
8125    73.9 bhp             190Nm@ 2000rpm    5.0  
8126      70 bhp        140Nm@ 1800-3000rpm    5.0  
8127      70 bhp        140Nm@ 1800-3000rpm    5.0  

[8128 rows x 13 columns]

In [3]:
#Deleting any duplicate row
df.drop_duplicates()

name  year  selling_price  km_driven  \
0                Maruti Swift Dzire VDI  2014         450000     145500   
1          Skoda Rapid 1.5 TDI Ambition  2014         370000     120000   
2              Honda City 2017-2020 EXi  2006         158000     140000   
3             Hyundai i20 Sportz Diesel  2010         225000     127000   
4                Maruti Swift VXI BSIII  2007         130000     120000   
...                                 ...   ...            ...        ...   
8121  Maruti Wagon R VXI BS IV with ABS  2013         260000      50000   
8122         Hyundai i20 Magna 1.4 CRDi  2014         475000      80000   
8123                  Hyundai i20 Magna  2013         320000     110000   
8124              Hyundai Verna CRDi SX  2007         135000     119000   
8125             Maruti Swift Dzire ZDi  2009         382000     120000   

        fuel seller_type transmission                 owner     mileage  \
0     Diesel  Individual       Manual           First Owner   23.4 kmpl   
1     Diesel  Individual       Manual          Second Owner  21.14 kmpl   
2     Petrol  Individual       Manual           Third Owner   17.7 kmpl   
3     Diesel  Individual       Manual           First Owner   23.0 kmpl   
4     Petrol  Individual       Manual           First Owner   16.1 kmpl   
...      ...         ...          ...                   ...         ...   
8121  Petrol  Individual       Manual          Second Owner   18.9 kmpl   
8122  Diesel  Individual       Manual          Second Owner  22.54 kmpl   
8123  Petrol  Individual       Manual           First Owner   18.5 kmpl   
8124  Diesel  Individual       Manual  Fourth & Above Owner   16.8 kmpl   
8125  Diesel  Individual       Manual           First Owner   19.3 kmpl   

       engine   max_power                     torque  seats  
0     1248 CC      74 bhp             190Nm@ 2000rpm    5.0  
1     1498 CC  103.52 bhp        250Nm@ 1500-2500rpm    5.0  
2     1497 CC      78 bhp      12.7@ 2,700(kgm@ rpm)    5.0  
3     1396 CC      90 bhp   22.4 kgm at 1750-2750rpm    5.0  
4     1298 CC    88.2 bhp      11.5@ 4,500(kgm@ rpm)    5.0  
...       ...         ...                        ...    ...  
8121   998 CC    67.1 bhp              90Nm@ 3500rpm    5.0  
8122  1396 CC   88.73 bhp      219.7Nm@ 1500-2750rpm    5.0  
8123  1197 CC   82.85 bhp           113.7Nm@ 4000rpm    5.0  
8124  1493 CC     110 bhp  24@ 1,900-2,750(kgm@ rpm)    5.0  
8125  1248 CC    73.9 bhp             190Nm@ 2000rpm    5.0  

[6926 rows x 13 columns]

In [4]:
df = df.drop_duplicates()
df

name  year  selling_price  km_driven  \
0                Maruti Swift Dzire VDI  2014         450000     145500   
1          Skoda Rapid 1.5 TDI Ambition  2014         370000     120000   
2              Honda City 2017-2020 EXi  2006         158000     140000   
3             Hyundai i20 Sportz Diesel  2010         225000     127000   
4                Maruti Swift VXI BSIII  2007         130000     120000   
...                                 ...   ...            ...        ...   
8121  Maruti Wagon R VXI BS IV with ABS  2013         260000      50000   
8122         Hyundai i20 Magna 1.4 CRDi  2014         475000      80000   
8123                  Hyundai i20 Magna  2013         320000     110000   
8124              Hyundai Verna CRDi SX  2007         135000     119000   
8125             Maruti Swift Dzire ZDi  2009         382000     120000   

        fuel seller_type transmission                 owner     mileage  \
0     Diesel  Individual       Manual           First Owner   23.4 kmpl   
1     Diesel  Individual       Manual          Second Owner  21.14 kmpl   
2     Petrol  Individual       Manual           Third Owner   17.7 kmpl   
3     Diesel  Individual       Manual           First Owner   23.0 kmpl   
4     Petrol  Individual       Manual           First Owner   16.1 kmpl   
...      ...         ...          ...                   ...         ...   
8121  Petrol  Individual       Manual          Second Owner   18.9 kmpl   
8122  Diesel  Individual       Manual          Second Owner  22.54 kmpl   
8123  Petrol  Individual       Manual           First Owner   18.5 kmpl   
8124  Diesel  Individual       Manual  Fourth & Above Owner   16.8 kmpl   
8125  Diesel  Individual       Manual           First Owner   19.3 kmpl   

       engine   max_power                     torque  seats  
0     1248 CC      74 bhp             190Nm@ 2000rpm    5.0  
1     1498 CC  103.52 bhp        250Nm@ 1500-2500rpm    5.0  
2     1497 CC      78 bhp      12.7@ 2,700(kgm@ rpm)    5.0  
3     1396 CC      90 bhp   22.4 kgm at 1750-2750rpm    5.0  
4     1298 CC    88.2 bhp      11.5@ 4,500(kgm@ rpm)    5.0  
...       ...         ...                        ...    ...  
8121   998 CC    67.1 bhp              90Nm@ 3500rpm    5.0  
8122  1396 CC   88.73 bhp      219.7Nm@ 1500-2750rpm    5.0  
8123  1197 CC   82.85 bhp           113.7Nm@ 4000rpm    5.0  
8124  1493 CC     110 bhp  24@ 1,900-2,750(kgm@ rpm)    5.0  
8125  1248 CC    73.9 bhp             190Nm@ 2000rpm    5.0  

[6926 rows x 13 columns]

We had 8128 rows and we only had 6926 left, that means we had 1202 duplicated rows.
Anyway, let´s see if there are some missing values.

In [5]:
#Checking the type of each column
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6926 entries, 0 to 8125
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   name           6926 non-null   object 
 1   year           6926 non-null   int64  
 2   selling_price  6926 non-null   int64  
 3   km_driven      6926 non-null   int64  
 4   fuel           6926 non-null   object 
 5   seller_type    6926 non-null   object 
 6   transmission   6926 non-null   object 
 7   owner          6926 non-null   object 
 8   mileage        6718 non-null   object 
 9   engine         6718 non-null   object 
 10  max_power      6721 non-null   object 
 11  torque         6717 non-null   object 
 12  seats          6718 non-null   float64
dtypes: float64(1), int64(3), object(9)
memory usage: 757.5+ KB


In [6]:
#Counting the missing values
df.isna().sum()

name               0
year               0
selling_price      0
km_driven          0
fuel               0
seller_type        0
transmission       0
owner              0
mileage          208
engine           208
max_power        205
torque           209
seats            208
dtype: int64

But first we notice that the selling price is actually given in rupees, we will convert this into dollars

In [7]:
#Converting Rupees to dollars 
Rupees_to_dollars = 0.012
df['selling_price'] = Rupees_to_dollars*df['selling_price']
df

name  year  selling_price  km_driven  \
0                Maruti Swift Dzire VDI  2014         5400.0     145500   
1          Skoda Rapid 1.5 TDI Ambition  2014         4440.0     120000   
2              Honda City 2017-2020 EXi  2006         1896.0     140000   
3             Hyundai i20 Sportz Diesel  2010         2700.0     127000   
4                Maruti Swift VXI BSIII  2007         1560.0     120000   
...                                 ...   ...            ...        ...   
8121  Maruti Wagon R VXI BS IV with ABS  2013         3120.0      50000   
8122         Hyundai i20 Magna 1.4 CRDi  2014         5700.0      80000   
8123                  Hyundai i20 Magna  2013         3840.0     110000   
8124              Hyundai Verna CRDi SX  2007         1620.0     119000   
8125             Maruti Swift Dzire ZDi  2009         4584.0     120000   

        fuel seller_type transmission                 owner     mileage  \
0     Diesel  Individual       Manual           First Owner   23.4 kmpl   
1     Diesel  Individual       Manual          Second Owner  21.14 kmpl   
2     Petrol  Individual       Manual           Third Owner   17.7 kmpl   
3     Diesel  Individual       Manual           First Owner   23.0 kmpl   
4     Petrol  Individual       Manual           First Owner   16.1 kmpl   
...      ...         ...          ...                   ...         ...   
8121  Petrol  Individual       Manual          Second Owner   18.9 kmpl   
8122  Diesel  Individual       Manual          Second Owner  22.54 kmpl   
8123  Petrol  Individual       Manual           First Owner   18.5 kmpl   
8124  Diesel  Individual       Manual  Fourth & Above Owner   16.8 kmpl   
8125  Diesel  Individual       Manual           First Owner   19.3 kmpl   

       engine   max_power                     torque  seats  
0     1248 CC      74 bhp             190Nm@ 2000rpm    5.0  
1     1498 CC  103.52 bhp        250Nm@ 1500-2500rpm    5.0  
2     1497 CC      78 bhp      12.7@ 2,700(kgm@ rpm)    5.0  
3     1396 CC      90 bhp   22.4 kgm at 1750-2750rpm    5.0  
4     1298 CC    88.2 bhp      11.5@ 4,500(kgm@ rpm)    5.0  
...       ...         ...                        ...    ...  
8121   998 CC    67.1 bhp              90Nm@ 3500rpm    5.0  
8122  1396 CC   88.73 bhp      219.7Nm@ 1500-2750rpm    5.0  
8123  1197 CC   82.85 bhp           113.7Nm@ 4000rpm    5.0  
8124  1493 CC     110 bhp  24@ 1,900-2,750(kgm@ rpm)    5.0  
8125  1248 CC    73.9 bhp             190Nm@ 2000rpm    5.0  

[6926 rows x 13 columns]

First, the Mileage column. 
We can simply drop all the missing values for this column, but instead we will replace the data, using the average mileage.
But before doing this we need to remove the signs '/kmplg'

In [8]:
#Droping characters
df['mileage'] = df['mileage'].str.strip('/kmplg')

We noticed that not every cell contains numeric values in this column so we transform them into numbers

In [9]:
df['mileage'] = pd.to_numeric(df['mileage'])
df

name  year  selling_price  km_driven  \
0                Maruti Swift Dzire VDI  2014         5400.0     145500   
1          Skoda Rapid 1.5 TDI Ambition  2014         4440.0     120000   
2              Honda City 2017-2020 EXi  2006         1896.0     140000   
3             Hyundai i20 Sportz Diesel  2010         2700.0     127000   
4                Maruti Swift VXI BSIII  2007         1560.0     120000   
...                                 ...   ...            ...        ...   
8121  Maruti Wagon R VXI BS IV with ABS  2013         3120.0      50000   
8122         Hyundai i20 Magna 1.4 CRDi  2014         5700.0      80000   
8123                  Hyundai i20 Magna  2013         3840.0     110000   
8124              Hyundai Verna CRDi SX  2007         1620.0     119000   
8125             Maruti Swift Dzire ZDi  2009         4584.0     120000   

        fuel seller_type transmission                 owner  mileage   engine  \
0     Diesel  Individual       Manual           First Owner    23.40  1248 CC   
1     Diesel  Individual       Manual          Second Owner    21.14  1498 CC   
2     Petrol  Individual       Manual           Third Owner    17.70  1497 CC   
3     Diesel  Individual       Manual           First Owner    23.00  1396 CC   
4     Petrol  Individual       Manual           First Owner    16.10  1298 CC   
...      ...         ...          ...                   ...      ...      ...   
8121  Petrol  Individual       Manual          Second Owner    18.90   998 CC   
8122  Diesel  Individual       Manual          Second Owner    22.54  1396 CC   
8123  Petrol  Individual       Manual           First Owner    18.50  1197 CC   
8124  Diesel  Individual       Manual  Fourth & Above Owner    16.80  1493 CC   
8125  Diesel  Individual       Manual           First Owner    19.30  1248 CC   

       max_power                     torque  seats  
0         74 bhp             190Nm@ 2000rpm    5.0  
1     103.52 bhp        250Nm@ 1500-2500rpm    5.0  
2         78 bhp      12.7@ 2,700(kgm@ rpm)    5.0  
3         90 bhp   22.4 kgm at 1750-2750rpm    5.0  
4       88.2 bhp      11.5@ 4,500(kgm@ rpm)    5.0  
...          ...                        ...    ...  
8121    67.1 bhp              90Nm@ 3500rpm    5.0  
8122   88.73 bhp      219.7Nm@ 1500-2750rpm    5.0  
8123   82.85 bhp           113.7Nm@ 4000rpm    5.0  
8124     110 bhp  24@ 1,900-2,750(kgm@ rpm)    5.0  
8125    73.9 bhp             190Nm@ 2000rpm    5.0  

[6926 rows x 13 columns]

And we replace the missing values with the mean mileage

In [10]:
avg_mil = df['mileage'].mean()
df['mileage'].fillna(value=avg_mil, inplace=True)
df

name  year  selling_price  km_driven  \
0                Maruti Swift Dzire VDI  2014         5400.0     145500   
1          Skoda Rapid 1.5 TDI Ambition  2014         4440.0     120000   
2              Honda City 2017-2020 EXi  2006         1896.0     140000   
3             Hyundai i20 Sportz Diesel  2010         2700.0     127000   
4                Maruti Swift VXI BSIII  2007         1560.0     120000   
...                                 ...   ...            ...        ...   
8121  Maruti Wagon R VXI BS IV with ABS  2013         3120.0      50000   
8122         Hyundai i20 Magna 1.4 CRDi  2014         5700.0      80000   
8123                  Hyundai i20 Magna  2013         3840.0     110000   
8124              Hyundai Verna CRDi SX  2007         1620.0     119000   
8125             Maruti Swift Dzire ZDi  2009         4584.0     120000   

        fuel seller_type transmission                 owner  mileage   engine  \
0     Diesel  Individual       Manual           First Owner    23.40  1248 CC   
1     Diesel  Individual       Manual          Second Owner    21.14  1498 CC   
2     Petrol  Individual       Manual           Third Owner    17.70  1497 CC   
3     Diesel  Individual       Manual           First Owner    23.00  1396 CC   
4     Petrol  Individual       Manual           First Owner    16.10  1298 CC   
...      ...         ...          ...                   ...      ...      ...   
8121  Petrol  Individual       Manual          Second Owner    18.90   998 CC   
8122  Diesel  Individual       Manual          Second Owner    22.54  1396 CC   
8123  Petrol  Individual       Manual           First Owner    18.50  1197 CC   
8124  Diesel  Individual       Manual  Fourth & Above Owner    16.80  1493 CC   
8125  Diesel  Individual       Manual           First Owner    19.30  1248 CC   

       max_power                     torque  seats  
0         74 bhp             190Nm@ 2000rpm    5.0  
1     103.52 bhp        250Nm@ 1500-2500rpm    5.0  
2         78 bhp      12.7@ 2,700(kgm@ rpm)    5.0  
3         90 bhp   22.4 kgm at 1750-2750rpm    5.0  
4       88.2 bhp      11.5@ 4,500(kgm@ rpm)    5.0  
...          ...                        ...    ...  
8121    67.1 bhp              90Nm@ 3500rpm    5.0  
8122   88.73 bhp      219.7Nm@ 1500-2750rpm    5.0  
8123   82.85 bhp           113.7Nm@ 4000rpm    5.0  
8124     110 bhp  24@ 1,900-2,750(kgm@ rpm)    5.0  
8125    73.9 bhp             190Nm@ 2000rpm    5.0  

[6926 rows x 13 columns]

And we make the same procedure as we did before with the next columns

In [11]:
df['engine'] = df['engine'].str.strip('C')
df['engine'] = pd.to_numeric(df['engine'])
df

name  year  selling_price  km_driven  \
0                Maruti Swift Dzire VDI  2014         5400.0     145500   
1          Skoda Rapid 1.5 TDI Ambition  2014         4440.0     120000   
2              Honda City 2017-2020 EXi  2006         1896.0     140000   
3             Hyundai i20 Sportz Diesel  2010         2700.0     127000   
4                Maruti Swift VXI BSIII  2007         1560.0     120000   
...                                 ...   ...            ...        ...   
8121  Maruti Wagon R VXI BS IV with ABS  2013         3120.0      50000   
8122         Hyundai i20 Magna 1.4 CRDi  2014         5700.0      80000   
8123                  Hyundai i20 Magna  2013         3840.0     110000   
8124              Hyundai Verna CRDi SX  2007         1620.0     119000   
8125             Maruti Swift Dzire ZDi  2009         4584.0     120000   

        fuel seller_type transmission                 owner  mileage  engine  \
0     Diesel  Individual       Manual           First Owner    23.40  1248.0   
1     Diesel  Individual       Manual          Second Owner    21.14  1498.0   
2     Petrol  Individual       Manual           Third Owner    17.70  1497.0   
3     Diesel  Individual       Manual           First Owner    23.00  1396.0   
4     Petrol  Individual       Manual           First Owner    16.10  1298.0   
...      ...         ...          ...                   ...      ...     ...   
8121  Petrol  Individual       Manual          Second Owner    18.90   998.0   
8122  Diesel  Individual       Manual          Second Owner    22.54  1396.0   
8123  Petrol  Individual       Manual           First Owner    18.50  1197.0   
8124  Diesel  Individual       Manual  Fourth & Above Owner    16.80  1493.0   
8125  Diesel  Individual       Manual           First Owner    19.30  1248.0   

       max_power                     torque  seats  
0         74 bhp             190Nm@ 2000rpm    5.0  
1     103.52 bhp        250Nm@ 1500-2500rpm    5.0  
2         78 bhp      12.7@ 2,700(kgm@ rpm)    5.0  
3         90 bhp   22.4 kgm at 1750-2750rpm    5.0  
4       88.2 bhp      11.5@ 4,500(kgm@ rpm)    5.0  
...          ...                        ...    ...  
8121    67.1 bhp              90Nm@ 3500rpm    5.0  
8122   88.73 bhp      219.7Nm@ 1500-2750rpm    5.0  
8123   82.85 bhp           113.7Nm@ 4000rpm    5.0  
8124     110 bhp  24@ 1,900-2,750(kgm@ rpm)    5.0  
8125    73.9 bhp             190Nm@ 2000rpm    5.0  

[6926 rows x 13 columns]

In [12]:
#Replacing missing values in engine with the mean value
mean_eng = df['engine'].mean()
df['engine'].fillna(value=mean_eng, inplace=True)
df

name  year  selling_price  km_driven  \
0                Maruti Swift Dzire VDI  2014         5400.0     145500   
1          Skoda Rapid 1.5 TDI Ambition  2014         4440.0     120000   
2              Honda City 2017-2020 EXi  2006         1896.0     140000   
3             Hyundai i20 Sportz Diesel  2010         2700.0     127000   
4                Maruti Swift VXI BSIII  2007         1560.0     120000   
...                                 ...   ...            ...        ...   
8121  Maruti Wagon R VXI BS IV with ABS  2013         3120.0      50000   
8122         Hyundai i20 Magna 1.4 CRDi  2014         5700.0      80000   
8123                  Hyundai i20 Magna  2013         3840.0     110000   
8124              Hyundai Verna CRDi SX  2007         1620.0     119000   
8125             Maruti Swift Dzire ZDi  2009         4584.0     120000   

        fuel seller_type transmission                 owner  mileage  engine  \
0     Diesel  Individual       Manual           First Owner    23.40  1248.0   
1     Diesel  Individual       Manual          Second Owner    21.14  1498.0   
2     Petrol  Individual       Manual           Third Owner    17.70  1497.0   
3     Diesel  Individual       Manual           First Owner    23.00  1396.0   
4     Petrol  Individual       Manual           First Owner    16.10  1298.0   
...      ...         ...          ...                   ...      ...     ...   
8121  Petrol  Individual       Manual          Second Owner    18.90   998.0   
8122  Diesel  Individual       Manual          Second Owner    22.54  1396.0   
8123  Petrol  Individual       Manual           First Owner    18.50  1197.0   
8124  Diesel  Individual       Manual  Fourth & Above Owner    16.80  1493.0   
8125  Diesel  Individual       Manual           First Owner    19.30  1248.0   

       max_power                     torque  seats  
0         74 bhp             190Nm@ 2000rpm    5.0  
1     103.52 bhp        250Nm@ 1500-2500rpm    5.0  
2         78 bhp      12.7@ 2,700(kgm@ rpm)    5.0  
3         90 bhp   22.4 kgm at 1750-2750rpm    5.0  
4       88.2 bhp      11.5@ 4,500(kgm@ rpm)    5.0  
...          ...                        ...    ...  
8121    67.1 bhp              90Nm@ 3500rpm    5.0  
8122   88.73 bhp      219.7Nm@ 1500-2750rpm    5.0  
8123   82.85 bhp           113.7Nm@ 4000rpm    5.0  
8124     110 bhp  24@ 1,900-2,750(kgm@ rpm)    5.0  
8125    73.9 bhp             190Nm@ 2000rpm    5.0  

[6926 rows x 13 columns]

In [13]:
df['max_power'] = df['max_power'].str.strip('bhp ') #Droping characters
df['max_power'] = pd.to_numeric(df['max_power']) #Converting to numerical 
mean_pow = df['max_power'].mean() #Calculating the mean value 
df['max_power'].fillna(value=mean_pow, inplace=True) #Replacing values
df

name  year  selling_price  km_driven  \
0                Maruti Swift Dzire VDI  2014         5400.0     145500   
1          Skoda Rapid 1.5 TDI Ambition  2014         4440.0     120000   
2              Honda City 2017-2020 EXi  2006         1896.0     140000   
3             Hyundai i20 Sportz Diesel  2010         2700.0     127000   
4                Maruti Swift VXI BSIII  2007         1560.0     120000   
...                                 ...   ...            ...        ...   
8121  Maruti Wagon R VXI BS IV with ABS  2013         3120.0      50000   
8122         Hyundai i20 Magna 1.4 CRDi  2014         5700.0      80000   
8123                  Hyundai i20 Magna  2013         3840.0     110000   
8124              Hyundai Verna CRDi SX  2007         1620.0     119000   
8125             Maruti Swift Dzire ZDi  2009         4584.0     120000   

        fuel seller_type transmission                 owner  mileage  engine  \
0     Diesel  Individual       Manual           First Owner    23.40  1248.0   
1     Diesel  Individual       Manual          Second Owner    21.14  1498.0   
2     Petrol  Individual       Manual           Third Owner    17.70  1497.0   
3     Diesel  Individual       Manual           First Owner    23.00  1396.0   
4     Petrol  Individual       Manual           First Owner    16.10  1298.0   
...      ...         ...          ...                   ...      ...     ...   
8121  Petrol  Individual       Manual          Second Owner    18.90   998.0   
8122  Diesel  Individual       Manual          Second Owner    22.54  1396.0   
8123  Petrol  Individual       Manual           First Owner    18.50  1197.0   
8124  Diesel  Individual       Manual  Fourth & Above Owner    16.80  1493.0   
8125  Diesel  Individual       Manual           First Owner    19.30  1248.0   

      max_power                     torque  seats  
0         74.00             190Nm@ 2000rpm    5.0  
1        103.52        250Nm@ 1500-2500rpm    5.0  
2         78.00      12.7@ 2,700(kgm@ rpm)    5.0  
3         90.00   22.4 kgm at 1750-2750rpm    5.0  
4         88.20      11.5@ 4,500(kgm@ rpm)    5.0  
...         ...                        ...    ...  
8121      67.10              90Nm@ 3500rpm    5.0  
8122      88.73      219.7Nm@ 1500-2750rpm    5.0  
8123      82.85           113.7Nm@ 4000rpm    5.0  
8124     110.00  24@ 1,900-2,750(kgm@ rpm)    5.0  
8125      73.90             190Nm@ 2000rpm    5.0  

[6926 rows x 13 columns]

The torque column contain many characters which are dispensable, in addition to the fact that it is very difficult to keep only the values that are useful to us, so we are going to get rid of this column, so as the seller type and seats columns 

In [14]:
df.drop(['seller_type', 'seats'], axis=1) #Droping columns

name  year  selling_price  km_driven  \
0                Maruti Swift Dzire VDI  2014         5400.0     145500   
1          Skoda Rapid 1.5 TDI Ambition  2014         4440.0     120000   
2              Honda City 2017-2020 EXi  2006         1896.0     140000   
3             Hyundai i20 Sportz Diesel  2010         2700.0     127000   
4                Maruti Swift VXI BSIII  2007         1560.0     120000   
...                                 ...   ...            ...        ...   
8121  Maruti Wagon R VXI BS IV with ABS  2013         3120.0      50000   
8122         Hyundai i20 Magna 1.4 CRDi  2014         5700.0      80000   
8123                  Hyundai i20 Magna  2013         3840.0     110000   
8124              Hyundai Verna CRDi SX  2007         1620.0     119000   
8125             Maruti Swift Dzire ZDi  2009         4584.0     120000   

        fuel transmission                 owner  mileage  engine  max_power  \
0     Diesel       Manual           First Owner    23.40  1248.0      74.00   
1     Diesel       Manual          Second Owner    21.14  1498.0     103.52   
2     Petrol       Manual           Third Owner    17.70  1497.0      78.00   
3     Diesel       Manual           First Owner    23.00  1396.0      90.00   
4     Petrol       Manual           First Owner    16.10  1298.0      88.20   
...      ...          ...                   ...      ...     ...        ...   
8121  Petrol       Manual          Second Owner    18.90   998.0      67.10   
8122  Diesel       Manual          Second Owner    22.54  1396.0      88.73   
8123  Petrol       Manual           First Owner    18.50  1197.0      82.85   
8124  Diesel       Manual  Fourth & Above Owner    16.80  1493.0     110.00   
8125  Diesel       Manual           First Owner    19.30  1248.0      73.90   

                         torque  
0                190Nm@ 2000rpm  
1           250Nm@ 1500-2500rpm  
2         12.7@ 2,700(kgm@ rpm)  
3      22.4 kgm at 1750-2750rpm  
4         11.5@ 4,500(kgm@ rpm)  
...                         ...  
8121              90Nm@ 3500rpm  
8122      219.7Nm@ 1500-2750rpm  
8123           113.7Nm@ 4000rpm  
8124  24@ 1,900-2,750(kgm@ rpm)  
8125             190Nm@ 2000rpm  

[6926 rows x 11 columns]

In [15]:
df = df.drop(['seller_type', 'seats'], axis=1)
df

name  year  selling_price  km_driven  \
0                Maruti Swift Dzire VDI  2014         5400.0     145500   
1          Skoda Rapid 1.5 TDI Ambition  2014         4440.0     120000   
2              Honda City 2017-2020 EXi  2006         1896.0     140000   
3             Hyundai i20 Sportz Diesel  2010         2700.0     127000   
4                Maruti Swift VXI BSIII  2007         1560.0     120000   
...                                 ...   ...            ...        ...   
8121  Maruti Wagon R VXI BS IV with ABS  2013         3120.0      50000   
8122         Hyundai i20 Magna 1.4 CRDi  2014         5700.0      80000   
8123                  Hyundai i20 Magna  2013         3840.0     110000   
8124              Hyundai Verna CRDi SX  2007         1620.0     119000   
8125             Maruti Swift Dzire ZDi  2009         4584.0     120000   

        fuel transmission                 owner  mileage  engine  max_power  \
0     Diesel       Manual           First Owner    23.40  1248.0      74.00   
1     Diesel       Manual          Second Owner    21.14  1498.0     103.52   
2     Petrol       Manual           Third Owner    17.70  1497.0      78.00   
3     Diesel       Manual           First Owner    23.00  1396.0      90.00   
4     Petrol       Manual           First Owner    16.10  1298.0      88.20   
...      ...          ...                   ...      ...     ...        ...   
8121  Petrol       Manual          Second Owner    18.90   998.0      67.10   
8122  Diesel       Manual          Second Owner    22.54  1396.0      88.73   
8123  Petrol       Manual           First Owner    18.50  1197.0      82.85   
8124  Diesel       Manual  Fourth & Above Owner    16.80  1493.0     110.00   
8125  Diesel       Manual           First Owner    19.30  1248.0      73.90   

                         torque  
0                190Nm@ 2000rpm  
1           250Nm@ 1500-2500rpm  
2         12.7@ 2,700(kgm@ rpm)  
3      22.4 kgm at 1750-2750rpm  
4         11.5@ 4,500(kgm@ rpm)  
...                         ...  
8121              90Nm@ 3500rpm  
8122      219.7Nm@ 1500-2750rpm  
8123           113.7Nm@ 4000rpm  
8124  24@ 1,900-2,750(kgm@ rpm)  
8125             190Nm@ 2000rpm  

[6926 rows x 11 columns]

The names in the first column are long and complicated. Let´s just leave the brand

First we see the unique names

In [16]:
names = df['name'].unique() 
names.sort()
names

array(['Ambassador CLASSIC 1500 DSL AC',
       'Ambassador Classic 2000 DSZ AC PS',
       'Ambassador Grand 1500 DSZ BSIII', ..., 'Volvo XC40 D4 R-Design',
       'Volvo XC60 Inscription D5 BSIV', 'Volvo XC90 T8 Excellence BSIV'],
      dtype=object)

And fortunately the first word of almost every row is the car´s brand, so we drop everything else

In [17]:
df['name'] = df['name'].apply(lambda x: x.split(' ')[0]) #Droping everything after the first word
df

name  year  selling_price  km_driven    fuel transmission  \
0      Maruti  2014         5400.0     145500  Diesel       Manual   
1       Skoda  2014         4440.0     120000  Diesel       Manual   
2       Honda  2006         1896.0     140000  Petrol       Manual   
3     Hyundai  2010         2700.0     127000  Diesel       Manual   
4      Maruti  2007         1560.0     120000  Petrol       Manual   
...       ...   ...            ...        ...     ...          ...   
8121   Maruti  2013         3120.0      50000  Petrol       Manual   
8122  Hyundai  2014         5700.0      80000  Diesel       Manual   
8123  Hyundai  2013         3840.0     110000  Petrol       Manual   
8124  Hyundai  2007         1620.0     119000  Diesel       Manual   
8125   Maruti  2009         4584.0     120000  Diesel       Manual   

                     owner  mileage  engine  max_power  \
0              First Owner    23.40  1248.0      74.00   
1             Second Owner    21.14  1498.0     103.52   
2              Third Owner    17.70  1497.0      78.00   
3              First Owner    23.00  1396.0      90.00   
4              First Owner    16.10  1298.0      88.20   
...                    ...      ...     ...        ...   
8121          Second Owner    18.90   998.0      67.10   
8122          Second Owner    22.54  1396.0      88.73   
8123           First Owner    18.50  1197.0      82.85   
8124  Fourth & Above Owner    16.80  1493.0     110.00   
8125           First Owner    19.30  1248.0      73.90   

                         torque  
0                190Nm@ 2000rpm  
1           250Nm@ 1500-2500rpm  
2         12.7@ 2,700(kgm@ rpm)  
3      22.4 kgm at 1750-2750rpm  
4         11.5@ 4,500(kgm@ rpm)  
...                         ...  
8121              90Nm@ 3500rpm  
8122      219.7Nm@ 1500-2750rpm  
8123           113.7Nm@ 4000rpm  
8124  24@ 1,900-2,750(kgm@ rpm)  
8125             190Nm@ 2000rpm  

[6926 rows x 11 columns]

And we replace certain values whose names are the model of the car and we replace them with the brand

In [18]:
names = df['name'].unique()
names.sort()
names

array(['Ambassador', 'Ashok', 'Audi', 'BMW', 'Chevrolet', 'Daewoo',
       'Datsun', 'Fiat', 'Force', 'Ford', 'Honda', 'Hyundai', 'Isuzu',
       'Jaguar', 'Jeep', 'Kia', 'Land', 'Lexus', 'MG', 'Mahindra',
       'Maruti', 'Mercedes-Benz', 'Mitsubishi', 'Nissan', 'Opel',
       'Peugeot', 'Renault', 'Skoda', 'Tata', 'Toyota', 'Volkswagen',
       'Volvo'], dtype=object)

In [19]:
df.replace({'name' : 
            {'Ambassador' : 'Hindustan', 'Land':'Land Rover'}  #Replacing some names for a better look
             })

name  year  selling_price  km_driven    fuel transmission  \
0      Maruti  2014         5400.0     145500  Diesel       Manual   
1       Skoda  2014         4440.0     120000  Diesel       Manual   
2       Honda  2006         1896.0     140000  Petrol       Manual   
3     Hyundai  2010         2700.0     127000  Diesel       Manual   
4      Maruti  2007         1560.0     120000  Petrol       Manual   
...       ...   ...            ...        ...     ...          ...   
8121   Maruti  2013         3120.0      50000  Petrol       Manual   
8122  Hyundai  2014         5700.0      80000  Diesel       Manual   
8123  Hyundai  2013         3840.0     110000  Petrol       Manual   
8124  Hyundai  2007         1620.0     119000  Diesel       Manual   
8125   Maruti  2009         4584.0     120000  Diesel       Manual   

                     owner  mileage  engine  max_power  \
0              First Owner    23.40  1248.0      74.00   
1             Second Owner    21.14  1498.0     103.52   
2              Third Owner    17.70  1497.0      78.00   
3              First Owner    23.00  1396.0      90.00   
4              First Owner    16.10  1298.0      88.20   
...                    ...      ...     ...        ...   
8121          Second Owner    18.90   998.0      67.10   
8122          Second Owner    22.54  1396.0      88.73   
8123           First Owner    18.50  1197.0      82.85   
8124  Fourth & Above Owner    16.80  1493.0     110.00   
8125           First Owner    19.30  1248.0      73.90   

                         torque  
0                190Nm@ 2000rpm  
1           250Nm@ 1500-2500rpm  
2         12.7@ 2,700(kgm@ rpm)  
3      22.4 kgm at 1750-2750rpm  
4         11.5@ 4,500(kgm@ rpm)  
...                         ...  
8121              90Nm@ 3500rpm  
8122      219.7Nm@ 1500-2750rpm  
8123           113.7Nm@ 4000rpm  
8124  24@ 1,900-2,750(kgm@ rpm)  
8125             190Nm@ 2000rpm  

[6926 rows x 11 columns]

For the year column, we will make some changes. 
Specifically we will change the year for the age of the car

First we change the year column to integers

In [20]:
df['year'] = df['year'].astype(str).astype(np.int64) #Converting to integers

And we crated a new column with the age of the car

In [21]:
Current_Year = 2023          
df['age'] = Current_Year - df['year']  #Calculating the age of the cars based on their year
df['age'] = df['age'].astype(np.int64)
df

name  year  selling_price  km_driven    fuel transmission  \
0      Maruti  2014         5400.0     145500  Diesel       Manual   
1       Skoda  2014         4440.0     120000  Diesel       Manual   
2       Honda  2006         1896.0     140000  Petrol       Manual   
3     Hyundai  2010         2700.0     127000  Diesel       Manual   
4      Maruti  2007         1560.0     120000  Petrol       Manual   
...       ...   ...            ...        ...     ...          ...   
8121   Maruti  2013         3120.0      50000  Petrol       Manual   
8122  Hyundai  2014         5700.0      80000  Diesel       Manual   
8123  Hyundai  2013         3840.0     110000  Petrol       Manual   
8124  Hyundai  2007         1620.0     119000  Diesel       Manual   
8125   Maruti  2009         4584.0     120000  Diesel       Manual   

                     owner  mileage  engine  max_power  \
0              First Owner    23.40  1248.0      74.00   
1             Second Owner    21.14  1498.0     103.52   
2              Third Owner    17.70  1497.0      78.00   
3              First Owner    23.00  1396.0      90.00   
4              First Owner    16.10  1298.0      88.20   
...                    ...      ...     ...        ...   
8121          Second Owner    18.90   998.0      67.10   
8122          Second Owner    22.54  1396.0      88.73   
8123           First Owner    18.50  1197.0      82.85   
8124  Fourth & Above Owner    16.80  1493.0     110.00   
8125           First Owner    19.30  1248.0      73.90   

                         torque  age  
0                190Nm@ 2000rpm    9  
1           250Nm@ 1500-2500rpm    9  
2         12.7@ 2,700(kgm@ rpm)   17  
3      22.4 kgm at 1750-2750rpm   13  
4         11.5@ 4,500(kgm@ rpm)   16  
...                         ...  ...  
8121              90Nm@ 3500rpm   10  
8122      219.7Nm@ 1500-2750rpm    9  
8123           113.7Nm@ 4000rpm   10  
8124  24@ 1,900-2,750(kgm@ rpm)   16  
8125             190Nm@ 2000rpm   14  

[6926 rows x 12 columns]

So that the column year will not be necessary

In [22]:
df = df.drop(['year'], axis=1)

We make a format for the fuel column since there are various types of fuel

In [23]:
fuel_type = df['fuel'].unique()
fuel_type.sort()
fuel_type

array(['CNG', 'Diesel', 'LPG', 'Petrol'], dtype=object)

We just replace CNG and LPG with gas

In [24]:
df.replace({'fuel' : 
            {'CNG' : 'Gas', 'LPG':'Gas'}
             })             #Replacing some types of fuel for an easier look

name  selling_price  km_driven    fuel transmission  \
0      Maruti         5400.0     145500  Diesel       Manual   
1       Skoda         4440.0     120000  Diesel       Manual   
2       Honda         1896.0     140000  Petrol       Manual   
3     Hyundai         2700.0     127000  Diesel       Manual   
4      Maruti         1560.0     120000  Petrol       Manual   
...       ...            ...        ...     ...          ...   
8121   Maruti         3120.0      50000  Petrol       Manual   
8122  Hyundai         5700.0      80000  Diesel       Manual   
8123  Hyundai         3840.0     110000  Petrol       Manual   
8124  Hyundai         1620.0     119000  Diesel       Manual   
8125   Maruti         4584.0     120000  Diesel       Manual   

                     owner  mileage  engine  max_power  \
0              First Owner    23.40  1248.0      74.00   
1             Second Owner    21.14  1498.0     103.52   
2              Third Owner    17.70  1497.0      78.00   
3              First Owner    23.00  1396.0      90.00   
4              First Owner    16.10  1298.0      88.20   
...                    ...      ...     ...        ...   
8121          Second Owner    18.90   998.0      67.10   
8122          Second Owner    22.54  1396.0      88.73   
8123           First Owner    18.50  1197.0      82.85   
8124  Fourth & Above Owner    16.80  1493.0     110.00   
8125           First Owner    19.30  1248.0      73.90   

                         torque  age  
0                190Nm@ 2000rpm    9  
1           250Nm@ 1500-2500rpm    9  
2         12.7@ 2,700(kgm@ rpm)   17  
3      22.4 kgm at 1750-2750rpm   13  
4         11.5@ 4,500(kgm@ rpm)   16  
...                         ...  ...  
8121              90Nm@ 3500rpm   10  
8122      219.7Nm@ 1500-2750rpm    9  
8123           113.7Nm@ 4000rpm   10  
8124  24@ 1,900-2,750(kgm@ rpm)   16  
8125             190Nm@ 2000rpm   14  

[6926 rows x 11 columns]

In [25]:
df = df.replace({'fuel' : 
            {'CNG' : 'Gas', 'LPG':'Gas'}
             })
df

name  selling_price  km_driven    fuel transmission  \
0      Maruti         5400.0     145500  Diesel       Manual   
1       Skoda         4440.0     120000  Diesel       Manual   
2       Honda         1896.0     140000  Petrol       Manual   
3     Hyundai         2700.0     127000  Diesel       Manual   
4      Maruti         1560.0     120000  Petrol       Manual   
...       ...            ...        ...     ...          ...   
8121   Maruti         3120.0      50000  Petrol       Manual   
8122  Hyundai         5700.0      80000  Diesel       Manual   
8123  Hyundai         3840.0     110000  Petrol       Manual   
8124  Hyundai         1620.0     119000  Diesel       Manual   
8125   Maruti         4584.0     120000  Diesel       Manual   

                     owner  mileage  engine  max_power  \
0              First Owner    23.40  1248.0      74.00   
1             Second Owner    21.14  1498.0     103.52   
2              Third Owner    17.70  1497.0      78.00   
3              First Owner    23.00  1396.0      90.00   
4              First Owner    16.10  1298.0      88.20   
...                    ...      ...     ...        ...   
8121          Second Owner    18.90   998.0      67.10   
8122          Second Owner    22.54  1396.0      88.73   
8123           First Owner    18.50  1197.0      82.85   
8124  Fourth & Above Owner    16.80  1493.0     110.00   
8125           First Owner    19.30  1248.0      73.90   

                         torque  age  
0                190Nm@ 2000rpm    9  
1           250Nm@ 1500-2500rpm    9  
2         12.7@ 2,700(kgm@ rpm)   17  
3      22.4 kgm at 1750-2750rpm   13  
4         11.5@ 4,500(kgm@ rpm)   16  
...                         ...  ...  
8121              90Nm@ 3500rpm   10  
8122      219.7Nm@ 1500-2750rpm    9  
8123           113.7Nm@ 4000rpm   10  
8124  24@ 1,900-2,750(kgm@ rpm)   16  
8125             190Nm@ 2000rpm   14  

[6926 rows x 11 columns]

In [26]:
transmission = df['transmission'].unique()
transmission.sort()
transmission  #Types of transmission

array(['Automatic', 'Manual'], dtype=object)

We don´t have problems with the transmission, since there are only two variables

So we continue with the number of owners

We would like to replace the words with the actual number of owners. First of all we see how many types are

In [27]:
owners = df['owner'].unique()
owners.sort()
owners #Types of owners

array(['First Owner', 'Fourth & Above Owner', 'Second Owner',
       'Test Drive Car', 'Third Owner'], dtype=object)

And we change the words with numbers

In [28]:
word_to_number = {'First Owner': 1, 'Second Owner': 2, 'Third Owner':3, 'Fourth & Above Owner':4, 'Test Drive Car':1}
df.replace({'owner': word_to_number}, inplace=True)
df  #Replacing words with number for a better treatment of data

name  selling_price  km_driven    fuel transmission  owner  mileage  \
0      Maruti         5400.0     145500  Diesel       Manual      1    23.40   
1       Skoda         4440.0     120000  Diesel       Manual      2    21.14   
2       Honda         1896.0     140000  Petrol       Manual      3    17.70   
3     Hyundai         2700.0     127000  Diesel       Manual      1    23.00   
4      Maruti         1560.0     120000  Petrol       Manual      1    16.10   
...       ...            ...        ...     ...          ...    ...      ...   
8121   Maruti         3120.0      50000  Petrol       Manual      2    18.90   
8122  Hyundai         5700.0      80000  Diesel       Manual      2    22.54   
8123  Hyundai         3840.0     110000  Petrol       Manual      1    18.50   
8124  Hyundai         1620.0     119000  Diesel       Manual      4    16.80   
8125   Maruti         4584.0     120000  Diesel       Manual      1    19.30   

      engine  max_power                     torque  age  
0     1248.0      74.00             190Nm@ 2000rpm    9  
1     1498.0     103.52        250Nm@ 1500-2500rpm    9  
2     1497.0      78.00      12.7@ 2,700(kgm@ rpm)   17  
3     1396.0      90.00   22.4 kgm at 1750-2750rpm   13  
4     1298.0      88.20      11.5@ 4,500(kgm@ rpm)   16  
...      ...        ...                        ...  ...  
8121   998.0      67.10              90Nm@ 3500rpm   10  
8122  1396.0      88.73      219.7Nm@ 1500-2750rpm    9  
8123  1197.0      82.85           113.7Nm@ 4000rpm   10  
8124  1493.0     110.00  24@ 1,900-2,750(kgm@ rpm)   16  
8125  1248.0      73.90             190Nm@ 2000rpm   14  

[6926 rows x 11 columns]

Finally all the columns seems right, excep for the torque, where there are still some missing  values. We simply delete all those 

In [29]:
df.dropna()

name  selling_price  km_driven    fuel transmission  owner  mileage  \
0      Maruti         5400.0     145500  Diesel       Manual      1    23.40   
1       Skoda         4440.0     120000  Diesel       Manual      2    21.14   
2       Honda         1896.0     140000  Petrol       Manual      3    17.70   
3     Hyundai         2700.0     127000  Diesel       Manual      1    23.00   
4      Maruti         1560.0     120000  Petrol       Manual      1    16.10   
...       ...            ...        ...     ...          ...    ...      ...   
8121   Maruti         3120.0      50000  Petrol       Manual      2    18.90   
8122  Hyundai         5700.0      80000  Diesel       Manual      2    22.54   
8123  Hyundai         3840.0     110000  Petrol       Manual      1    18.50   
8124  Hyundai         1620.0     119000  Diesel       Manual      4    16.80   
8125   Maruti         4584.0     120000  Diesel       Manual      1    19.30   

      engine  max_power                     torque  age  
0     1248.0      74.00             190Nm@ 2000rpm    9  
1     1498.0     103.52        250Nm@ 1500-2500rpm    9  
2     1497.0      78.00      12.7@ 2,700(kgm@ rpm)   17  
3     1396.0      90.00   22.4 kgm at 1750-2750rpm   13  
4     1298.0      88.20      11.5@ 4,500(kgm@ rpm)   16  
...      ...        ...                        ...  ...  
8121   998.0      67.10              90Nm@ 3500rpm   10  
8122  1396.0      88.73      219.7Nm@ 1500-2750rpm    9  
8123  1197.0      82.85           113.7Nm@ 4000rpm   10  
8124  1493.0     110.00  24@ 1,900-2,750(kgm@ rpm)   16  
8125  1248.0      73.90             190Nm@ 2000rpm   14  

[6717 rows x 11 columns]

In [30]:
df = df.dropna()
df

name  selling_price  km_driven    fuel transmission  owner  mileage  \
0      Maruti         5400.0     145500  Diesel       Manual      1    23.40   
1       Skoda         4440.0     120000  Diesel       Manual      2    21.14   
2       Honda         1896.0     140000  Petrol       Manual      3    17.70   
3     Hyundai         2700.0     127000  Diesel       Manual      1    23.00   
4      Maruti         1560.0     120000  Petrol       Manual      1    16.10   
...       ...            ...        ...     ...          ...    ...      ...   
8121   Maruti         3120.0      50000  Petrol       Manual      2    18.90   
8122  Hyundai         5700.0      80000  Diesel       Manual      2    22.54   
8123  Hyundai         3840.0     110000  Petrol       Manual      1    18.50   
8124  Hyundai         1620.0     119000  Diesel       Manual      4    16.80   
8125   Maruti         4584.0     120000  Diesel       Manual      1    19.30   

      engine  max_power                     torque  age  
0     1248.0      74.00             190Nm@ 2000rpm    9  
1     1498.0     103.52        250Nm@ 1500-2500rpm    9  
2     1497.0      78.00      12.7@ 2,700(kgm@ rpm)   17  
3     1396.0      90.00   22.4 kgm at 1750-2750rpm   13  
4     1298.0      88.20      11.5@ 4,500(kgm@ rpm)   16  
...      ...        ...                        ...  ...  
8121   998.0      67.10              90Nm@ 3500rpm   10  
8122  1396.0      88.73      219.7Nm@ 1500-2750rpm    9  
8123  1197.0      82.85           113.7Nm@ 4000rpm   10  
8124  1493.0     110.00  24@ 1,900-2,750(kgm@ rpm)   16  
8125  1248.0      73.90             190Nm@ 2000rpm   14  

[6717 rows x 11 columns]

In [31]:
df.isna().sum() #Is there any missing value?

name             0
selling_price    0
km_driven        0
fuel             0
transmission     0
owner            0
mileage          0
engine           0
max_power        0
torque           0
age              0
dtype: int64

Now we clean the values of torque. We notice that there are more information than just the torque, so we drop all these

In [32]:
df['torque'] = df['torque'].astype("string") #Converting to str

In [33]:
df['torque'] = df['torque'].apply(lambda x: x.split(' ')[0]) #Splitting values before the space
df

name  selling_price  km_driven    fuel transmission  owner  mileage  \
0      Maruti         5400.0     145500  Diesel       Manual      1    23.40   
1       Skoda         4440.0     120000  Diesel       Manual      2    21.14   
2       Honda         1896.0     140000  Petrol       Manual      3    17.70   
3     Hyundai         2700.0     127000  Diesel       Manual      1    23.00   
4      Maruti         1560.0     120000  Petrol       Manual      1    16.10   
...       ...            ...        ...     ...          ...    ...      ...   
8121   Maruti         3120.0      50000  Petrol       Manual      2    18.90   
8122  Hyundai         5700.0      80000  Diesel       Manual      2    22.54   
8123  Hyundai         3840.0     110000  Petrol       Manual      1    18.50   
8124  Hyundai         1620.0     119000  Diesel       Manual      4    16.80   
8125   Maruti         4584.0     120000  Diesel       Manual      1    19.30   

      engine  max_power    torque  age  
0     1248.0      74.00    190Nm@    9  
1     1498.0     103.52    250Nm@    9  
2     1497.0      78.00     12.7@   17  
3     1396.0      90.00      22.4   13  
4     1298.0      88.20     11.5@   16  
...      ...        ...       ...  ...  
8121   998.0      67.10     90Nm@   10  
8122  1396.0      88.73  219.7Nm@    9  
8123  1197.0      82.85  113.7Nm@   10  
8124  1493.0     110.00       24@   16  
8125  1248.0      73.90    190Nm@   14  

[6717 rows x 11 columns]

And finally delete the signs Nm@

In [34]:
df['torque'].str.strip('Nm@') #Deleting characters

0         190
1         250
2        12.7
3        22.4
4        11.5
        ...  
8121       90
8122    219.7
8123    113.7
8124       24
8125      190
Name: torque, Length: 6717, dtype: object

In [35]:
df['torque'] = df['torque'].str.strip('Nnmkg@')
df

name  selling_price  km_driven    fuel transmission  owner  mileage  \
0      Maruti         5400.0     145500  Diesel       Manual      1    23.40   
1       Skoda         4440.0     120000  Diesel       Manual      2    21.14   
2       Honda         1896.0     140000  Petrol       Manual      3    17.70   
3     Hyundai         2700.0     127000  Diesel       Manual      1    23.00   
4      Maruti         1560.0     120000  Petrol       Manual      1    16.10   
...       ...            ...        ...     ...          ...    ...      ...   
8121   Maruti         3120.0      50000  Petrol       Manual      2    18.90   
8122  Hyundai         5700.0      80000  Diesel       Manual      2    22.54   
8123  Hyundai         3840.0     110000  Petrol       Manual      1    18.50   
8124  Hyundai         1620.0     119000  Diesel       Manual      4    16.80   
8125   Maruti         4584.0     120000  Diesel       Manual      1    19.30   

      engine  max_power torque  age  
0     1248.0      74.00    190    9  
1     1498.0     103.52    250    9  
2     1497.0      78.00   12.7   17  
3     1396.0      90.00   22.4   13  
4     1298.0      88.20   11.5   16  
...      ...        ...    ...  ...  
8121   998.0      67.10     90   10  
8122  1396.0      88.73  219.7    9  
8123  1197.0      82.85  113.7   10  
8124  1493.0     110.00     24   16  
8125  1248.0      73.90    190   14  

[6717 rows x 11 columns]

In [36]:
torque_b = df[df['torque'] == '380Nm(38.7kgm)'].index
torque_b   #Another value with characters

Index([778], dtype='int64')

In [37]:
df.drop(torque_b, inplace=True) #Droping it

In [38]:
df['torque'] = df['torque'].apply(lambda x: x.split('(')[0])
df
#Droping values before the ( sign

name  selling_price  km_driven    fuel transmission  owner  mileage  \
0      Maruti         5400.0     145500  Diesel       Manual      1    23.40   
1       Skoda         4440.0     120000  Diesel       Manual      2    21.14   
2       Honda         1896.0     140000  Petrol       Manual      3    17.70   
3     Hyundai         2700.0     127000  Diesel       Manual      1    23.00   
4      Maruti         1560.0     120000  Petrol       Manual      1    16.10   
...       ...            ...        ...     ...          ...    ...      ...   
8121   Maruti         3120.0      50000  Petrol       Manual      2    18.90   
8122  Hyundai         5700.0      80000  Diesel       Manual      2    22.54   
8123  Hyundai         3840.0     110000  Petrol       Manual      1    18.50   
8124  Hyundai         1620.0     119000  Diesel       Manual      4    16.80   
8125   Maruti         4584.0     120000  Diesel       Manual      1    19.30   

      engine  max_power torque  age  
0     1248.0      74.00    190    9  
1     1498.0     103.52    250    9  
2     1497.0      78.00   12.7   17  
3     1396.0      90.00   22.4   13  
4     1298.0      88.20   11.5   16  
...      ...        ...    ...  ...  
8121   998.0      67.10     90   10  
8122  1396.0      88.73  219.7    9  
8123  1197.0      82.85  113.7   10  
8124  1493.0     110.00     24   16  
8125  1248.0      73.90    190   14  

[6716 rows x 11 columns]

In [39]:
df['torque'] = df['torque'].str.strip('NM') #Deleting characters

In [40]:
df['torque'] = df['torque'].values.astype(float) #Converting to float

This is our clean dataset

In [41]:
df

name  selling_price  km_driven    fuel transmission  owner  mileage  \
0      Maruti         5400.0     145500  Diesel       Manual      1    23.40   
1       Skoda         4440.0     120000  Diesel       Manual      2    21.14   
2       Honda         1896.0     140000  Petrol       Manual      3    17.70   
3     Hyundai         2700.0     127000  Diesel       Manual      1    23.00   
4      Maruti         1560.0     120000  Petrol       Manual      1    16.10   
...       ...            ...        ...     ...          ...    ...      ...   
8121   Maruti         3120.0      50000  Petrol       Manual      2    18.90   
8122  Hyundai         5700.0      80000  Diesel       Manual      2    22.54   
8123  Hyundai         3840.0     110000  Petrol       Manual      1    18.50   
8124  Hyundai         1620.0     119000  Diesel       Manual      4    16.80   
8125   Maruti         4584.0     120000  Diesel       Manual      1    19.30   

      engine  max_power  torque  age  
0     1248.0      74.00   190.0    9  
1     1498.0     103.52   250.0    9  
2     1497.0      78.00    12.7   17  
3     1396.0      90.00    22.4   13  
4     1298.0      88.20    11.5   16  
...      ...        ...     ...  ...  
8121   998.0      67.10    90.0   10  
8122  1396.0      88.73   219.7    9  
8123  1197.0      82.85   113.7   10  
8124  1493.0     110.00    24.0   16  
8125  1248.0      73.90   190.0   14  

[6716 rows x 11 columns]

In [42]:
df.to_csv('cleaned_car_data.csv', index = False)  #Converting to a csv for use it in another Notebooks